src/main/java/org/sonatype/goodies/packageurl/PackageUrlBuilder.java

```diff
@@ -158,19 +172,39 @@ PackageUrl buildAndValidate(final boolean validate) {
     // FIXME: need to have some per-type transformation; which is unfortunate but spec requires some special handling per-type
     // FIXME: various type-specific transformation required by specification; very problematic
     // FIXME: https://github.com/package-url/purl-spec/issues/38
-    switch (type) {
-      case "github":
-      case "bitbucket":
-        name = MoreStrings.lowerCase(name);
-        namespace = MoreStrings.lowerCase(namespace);
-        break;
-
-      case "pypi":
-        name = name.replace('_', '-');
-        name = MoreStrings.lowerCase(name);
-        break;
+    List<String> correctedNamespace = namespace;
+    String correctedName = name;
+    if (correct) {
+      switch (type) {
+        case "github":
+        case "bitbucket":
+          correctedNamespace = MoreStrings.lowerCase(namespace);
+          correctedName = MoreStrings.lowerCase(name);
+          break;
+
+        case "pypi":
+          correctedName = name.replace('_', '-');
+          correctedName = MoreStrings.lowerCase(correctedName);
+          break;
+      }
+    }
+
+    SortedMap<String, String> correctedQualifiers = null;
+    if (qualifiers != null) {
+      correctedQualifiers = new TreeMap<>();
+      for (Entry<String, String> entry : qualifiers.entrySet()) {
+        String key = MoreStrings.lowerCase(entry.getKey());
```

Comment:
Yes, all of our qualifiers are in fact lower case, so for DI-1388, this is a moot point.
Originally I had "as-is" mode lower-case these, but @jdillon would like to keep as close to the spec as possible, so I reverted the change.

Reply:
> Originally I had "as-is" mode lower-case these, but @jdillon would like to keep as close to the spec as possible, so I reverted the change.

I'm confused. It looks to me like it still lowerCases entry.getKey() and puts it in correctedQualifiers as long as the value's not empty.

Reply:
FTR I think "asis" may be misleading, I think the intent with this flag is to just avoid doing the additional per-type transformations (like for pypi lowercasing and translating '_' to '-' or for npm lowercasing) etc.

Otherwise I think we should try to behave as close as possible to the spec.

Reply:
maybe we should use a different term than as-is?
* `transform`
* `transformPypiAndBitbucketAndGithub` -- joking 😝
* `strictlyFollowSpec` - as in, strictly follow the spec
* `naive` - as in, naively assume input data is good and shouldn't be transformed

Reply:
agree I think we should not use "asis", .... "enable/disable silly transformation quirks" or something to that effect

Reply:
@jdillon what term would you like me to use instead of "as-is"? (or "!as-is")?

Reply:
How about `preserveNaming`, or even `keepNaming`?

Reply:
How about this?

boolean typeSpecificTransformationQuirks = true;
public PackageUrlBuilder typeSpecificTransformationQuirks(boolean enable)

Reply:
:-D

Reply:
Went with "typeSpecificTransformations"